File: ios/Classes/Methods/PluginMethods.swift
import Flutter
import TwilioConversationsClient

class PluginMethods: NSObject, TWCONPluginApi {
    let TAG = "PluginMethods"

    func debugEnableNative(
        _ enableNative: NSNumber,
        enableSdk: NSNumber,
        error: AutoreleasingUnsafeMutablePointer<FlutterError?>) {
        SwiftTwilioConversationsPlugin.nativeDebug = enableNative.boolValue
        if enableSdk.boolValue {
            TwilioConversationsClient.setLogLevel(TCHLogLevel.debug)
        } else {
            TwilioConversationsClient.setLogLevel(TCHLogLevel.warning)
        }
    }

    // Naming a by product of pigeon generation. This creates a conversations client, not a JWT token.
    func createJwtToken(
        _ jwtToken: String?,
        properties: TWCONPropertiesData?,
        completion: @escaping (TWCONConversationClientData?, FlutterError?) -> Void) {
        guard let jwtToken = jwtToken else {
            return completion(
                nil,
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing 'token' parameter",
                    details: nil))
        }

        guard let properties = properties else {
            return completion(
                nil,
                FlutterError(
                    code: "MissingParameterException",
                    message: "Missing 'properties' parameter",
                    details: nil))
        }
        debug("create => jwtToken: \(jwtToken)")

        let clientProperties = TwilioConversationsClientProperties()
        clientProperties.region = properties.region ?? "us1"

        SwiftTwilioConversationsPlugin.clientListener = ClientListener()

        TwilioConversationsClient.conversationsClient(
            withToken: jwtToken,
            properties: clientProperties,
            delegate: SwiftTwilioConversationsPlugin.clientListener,
            completion: { (result: TCHResult, conversationsClient: TwilioConversationsClient?) in
                if result.isSuccessful {
                    let myIdentity = conversationsClient?.user?.identity ?? "unknown"
                    self.debug("create => onSuccess - myIdentity: '\(myIdentity)'")
                    conversationsClient?.delegate = SwiftTwilioConversationsPlugin.clientListener
                    SwiftTwilioConversationsPlugin.instance?.client = conversationsClient
                    let clientData = Mapper.conversationsClientToPigeon(conversationsClient)
                    completion(clientData, nil)
                } else {
                    self.debug("create => onError: \(String(describing: result.error))")
                    completion(
                        nil,
                        FlutterError(
                            code: "TwilioException",
                            message: "\(result.error?.code)|Error creating client, Error: "
                            + "\(result.error.debugDescription)",
                            details: nil))
                }
        })
    }

    private func debug(_ msg: String) {
        SwiftTwilioConversationsPlugin.debug("\(TAG)::\(msg)")
    }
}

File: android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/PluginMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.conversations.ConversationsClient
import com.twilio.util.ErrorInfo
import twilio.flutter.twilio_conversations.Api
import twilio.flutter.twilio_conversations.Mapper
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import twilio.flutter.twilio_conversations.exceptions.TwilioException
import twilio.flutter.twilio_conversations.listeners.ClientListener
import twilio.flutter.twilio_conversations.listeners.SafeCallbackListener

class PluginMethods : Api.PluginApi {
    private val TAG = "PluginMethods"

    override fun debug(enableNative: Boolean, enableSdk: Boolean) {
        if (enableSdk) {
            ConversationsClient.setLogLevel(ConversationsClient.LogLevel.DEBUG)
        } else {
            ConversationsClient.setLogLevel(ConversationsClient.LogLevel.ERROR)
        }

        TwilioConversationsPlugin.nativeDebug = enableNative
        return
    }

    override fun create(jwtToken: String, properties: Api.PropertiesData, result: Api.Result<Api.ConversationClientData>) {
        debug("create => jwtToken: $jwtToken")

        try {
            val props = ConversationsClient.Properties.newBuilder()
                .setRegion(properties.region)
                .createProperties()

            ConversationsClient.create(
                TwilioConversationsPlugin.applicationContext,
                jwtToken,
                props,
                object :
                    SafeCallbackListener<ConversationsClient> {
                    override fun onSafeSuccess(item: ConversationsClient) {
                        debug("create => onSuccess - myIdentity: '${item.myUser?.identity ?: "unknown"}'")
                        TwilioConversationsPlugin.client = item
                        TwilioConversationsPlugin.clientListener = ClientListener()
                        item.addListener(TwilioConversationsPlugin.clientListener!!)
                        val clientMap = Mapper.conversationsClientToPigeon(item)
                        result.success(clientMap)
                    }

                    override fun onError(errorInfo: ErrorInfo) {
                        debug("create => onError: ${errorInfo.message}")
                        result.error(TwilioException(errorInfo.code, errorInfo.message))
                    }
                })
        } catch (e: Exception) {
            debug("create => onError: ${e.message}")
            result.error(TwilioException(-1, e.message ?: "Unknown error"))
        }
    }

    fun debug(message: String) {
        TwilioConversationsPlugin.debug("$TAG::$message")
    }
}
